crates/sui-types/src/temporary_store.rs

```diff
@@ -729,46 +719,61 @@ impl<S> TemporaryStore<S> {
-            .clone();
-        trace!(?gas_object_id, "Obtained gas object");
-        if let Err(err) = self.charge_gas_for_storage_changes(sender, gas_status, &mut gas_object) {
-            // If `result` is already `Err`, we basically have two errors at the same time.
-            // Users should be generally more interested in the actual execution error, so we
-            // let that shadow the out of gas error. Also in this case, we don't need to reset
-            // the `temporary_store` because `charge_gas_for_storage_changes` won't mutate
-            // `temporary_store` if gas charge failed.
-            //
-            // If `result` is `Ok`, now we failed when charging gas, we have to reset
-            // the `temporary_store` to eliminate all effects caused by the execution,
-            // and re-ensure all mutable objects' versions are incremented.
-            if result.is_ok() {
-                self.reset();
-                let gas_object_id = match self.smash_gas(sender, gas) {
-                    Ok(obj_ref) => obj_ref.0,
-                    Err(_) => gas[0].0, // this cannot fail, but we use gas[0] anyway
-                };
-                self.ensure_active_inputs_mutated(sender, &gas_object_id);
-                *result = Err(err);
+        // at this point, we have done some charging for computation, but have not yet set the storage rebate or storage gas units
+        assert!(gas_status.storage_rebate() == 0);
+        assert!(gas_status.storage_gas_units() == 0);
+
+        if execution_result.is_err() {
+            // Tx execution aborted--need to dump writes, deletes, etc before charging storage gas
+            self.reset(sender, gas, gas_status);
+        }
+
+        if let Err(err) = self
+            .charge_gas_for_storage_changes(sender, gas_status, gas_object_id)
+            .and_then(|total_bytes_written_deleted| {
+                gas_status.charge_computation_gas_for_storage_mutation(total_bytes_written_deleted)
+            })
+        {
+            // Ran out of gas while charging for storage changes. reset store, now at state just after gas smashing
+            self.reset(sender, gas, gas_status);
+
+            // charge for storage again. This will now account only for the storage cost of gas coins
+            if self
+                .charge_gas_for_storage_changes(sender, gas_status, gas_object_id)
+                .and_then(|total_bytes_written_deleted| {
+                    gas_status
+                        .charge_computation_gas_for_storage_mutation(total_bytes_written_deleted)
+                })
+                .is_err()
+            {
+                // TODO: this shouldn't happen, because we should check that the budget is enough to cover the storage costs of gas coins at signing time
+                // perhaps that check isn't there?
+                trace!("out of gas while charging for gas smashing")
```

Comment:
I think eventually we want an `assert(false)` here + to do the corresponding check when signing to ensure that it won't fail. But a lot of tests failed when I tried to do `assert(false)` here, hence the print.

Reply:
good to know, I'll double check